src/mapping/NearestNeighborGradientMapping.hpp

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include "logging/Logger.hpp"
+#include "mapping/NearestNeighborBaseMapping.hpp"
+
+namespace precice {
+namespace mapping {
+
+/// Mapping using nearest neighboring vertices and their local gradient values.
+class NearestNeighborGradientMapping : public NearestNeighborBaseMapping {
+public:
+  /**
+   * @brief Constructor.
+   *
+   * @param[in] constraint Specifies mapping to be consistent or conservative.
+   * @param[in] dimensions Dimensionality of the meshes
+   */
+  NearestNeighborGradientMapping(Constraint constraint, int dimensions);
+
+  /// Destructor, empty.
+  virtual ~NearestNeighborGradientMapping() {}
+  
+  virtual double mapAt(int mapInputIndex, int vertex, const Eigen::VectorXd &inputValues, const Eigen::MatrixXd &gradientValues) override ;
```

Comment:
As a general remark: This PR is getting quite big already, which makes it a bit hard to review. I would propose to open a separate PR to the main repo including the NN refactoring only (and probably explaining why you want to do it).

The general layout of the refactoring looks good. The only thing to discuss is whether it is a good idea to use a virtual function on such a low level: `mapAt` is called on every vertex in every iteration, you can read in [this discussion](https://softwareengineering.stackexchange.com/questions/191637/in-c-why-and-how-are-virtual-functions-slower), why virtual functions harm performance. A better approach would probably be to to put the for loop over all vertices into the virtual function so that it is only called once per mapping.

However, you can keep it for the moment this way and we can discuss this in the refactoring PR.